File: Alutec AUDI/HidriaVisionWinForms/XMLSettings/XMLSettings/XMLSettingsManager.cs
using System;
using System.Dynamic;
using System.Xml.Linq;
using System.Linq;
using System.Collections.Generic;
using System.Globalization;
using System.IO;

namespace XMLSettings
{
    /// <summary>
    /// Represents XML settings manager.
    /// </summary>
    public class XMLSettingsManager
    {
        #region Private fields
        private XElement XMLRoot;
        private string XMLPath;
        private object ObjLock;
        private string FileName;
        #endregion

        /// <summary>
        /// Constructs an object of type XMLSettingsManager.
        /// </summary>
        /// <param name="folderPath">Folder path.</param>
        /// <param name="fileName">File name.</param>
        /// <exception cref="XmlException"></exception>
        /// <exception cref="FileNotFoundException"></exception>
        /// <exception cref="SecurityException"></exception>
        /// <exception cref="ArgumentNullException"></exception>
        public XMLSettingsManager(string folderPath, string fileName)
        {
            // Check if folder path is null or empty
            if (string.IsNullOrEmpty(folderPath))
                throw new ArgumentException("Folder path is null or empty.");

            // Check if file name is null or empty
            if (string.IsNullOrEmpty(fileName))
                throw new ArgumentException("File name is null or empty.");

            // Parse an XML file and initialize lock
            XMLRoot = XElement.Load(Path.GetFullPath(string.Format(@"{0}\{1}", folderPath, fileName)));
            FileName = fileName;
            XMLPath = folderPath;
            ObjLock = new object();
        }

        #region Public methods
        /// <summary>
        /// Gets the specified element.
        /// </summary>
        /// <param name="segmentName">Segment name.</param>
        /// <param name="attribute">Segment attribute.</param>
        /// <param name="name">Element name.</param>
        /// <returns>Returns value of a specified element.</returns>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentException"></exception>
        public object GetElement(string segmentName, string segmentAttribute, string name) 
        {
            // Check if segment name is null or empty
            if (string.IsNullOrEmpty(segmentName))
                throw new ArgumentException("Segment name is null or empty.");

            // Check if segment attribute is null or empty
            if (string.IsNullOrEmpty(segmentAttribute))
                throw new ArgumentException("Segment attribute is null or empty.");

            // Check if name is null or empty
            if (string.IsNullOrEmpty(name))
                throw new ArgumentException("Name is null or empty.");

            // Lock and return element's value
            lock (ObjLock)
            {
                IEnumerable<XElement> SegmentElements;
                XElement SegmentElement;
                Type ElementType;

                // Get segment elements
                SegmentElements = GetSegmentElements(segmentName, segmentAttribute);

                // Check that only one element with a given name exists
                if (SegmentElements.Where(x => x.Name == name).Count() != 1)
                    throw new InvalidOperationException("There are no elements or there are multiple elements with the same name.");

                // Get segment element
                SegmentElement = GetSegmentElement(SegmentElements, name);

                // Check if type has been specified inside the element
                if (SegmentElement.Attribute("Type") == null)
                    throw new InvalidOperationException("Type not specified.");

                // Get element type
                ElementType = GetElementType(SegmentElement);

                // Return element's value
                return Convert.ChangeType(SegmentElement.Value, ElementType, CultureInfo.InvariantCulture);
            }
        }

        /// <summary>
        /// Sets the specified element.
        /// </summary>
        /// <param name="segmentName">Segment name.</param>
        /// <param name="attribute">Segment attribute.</param>
        /// <param name="name">Element name.</param>
        /// <param name="value">Element value.</param>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentException"></exception>
        public void SetElement(string segmentName, string segmentAttribute, string name, object value)
        {
            // Check if segment name is null or empty
            if (string.IsNullOrEmpty(segmentName))
                throw new ArgumentException("Segment name is null or empty.");

            // Check if segment attribute is null or empty
            if (string.IsNullOrEmpty(segmentAttribute))
                throw new ArgumentException("Segment attribute is null or empty.");

            // Check if name is null or empty
            if (string.IsNullOrEmpty(name))
                throw new ArgumentException("Name is null or empty.");

            // Check if value is null
            if (value == null)
                throw new ArgumentException("Value is null.");

            // Lock and set element's value
            lock (ObjLock)
            {
                IEnumerable<XElement> SegmentElements;
                XElement SegmentElement;
                Type ElementType;

                // Get segment elements
                SegmentElements = GetSegmentElements(segmentName, segmentAttribute);

                // Check that only one element with a given name exists
                if (SegmentElements.Where(x => x.Name == name).Count() != 1)
                    throw new InvalidOperationException("There are no elements or there are multiple elements with the same name.");

                // Get segment element
                SegmentElement = GetSegmentElement(SegmentElements, name);

                // Check if type has been specified inside the element
                if (SegmentElement.Attribute("Type") == null)
                    throw new InvalidOperationException("Type not specified.");

                // Get element type
                ElementType = GetElementType(SegmentElement);

                // Check if value type is valid
                if (value.GetType() != ElementType)
                    throw new InvalidOperationException("Value type is invalid.");

                // Convert value to string 
                SegmentElement.Value = Convert.ToString(value, CultureInfo.InvariantCulture);

                // Save the xml
                XMLRoot.Save(Path.Combine(XMLPath, FileName));
            }
        }

        /// <summary>
        /// Gets the specified segment.
        /// </summary>
        /// <param name="segmentName">Segment name.</param>
        /// <param name="segmentAttribute">Segment attribute.</param>
        /// <returns>Returns the specified segment.</returns>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentException"></exception>
        public object GetSegment(string segmentName, string segmentAttribute)
        {
            // Check if segment name is null or empty
            if (string.IsNullOrEmpty(segmentName))
                throw new ArgumentException("Segment name is null or empty.");

            // Check if segment attribute is null or empty
            if (string.IsNullOrEmpty(segmentAttribute))
                throw new ArgumentException("Segment attribute is null or empty.");

            // Lock and get segment
            lock (ObjLock)
            {
                dynamic Segment = new ExpandoObject();
                IEnumerable<XElement> SegmentElements;
                IDictionary<string, object> ExpandoDict = Segment as IDictionary<string, object>;

                // Get segment elements
                SegmentElements = GetSegmentElements(segmentName, segmentAttribute);

                // Go through the elements
                foreach (XElement segmentElement in SegmentElements)
                {
                    Type ElementType;

                    // Check that only one element with a given name exists
                    if (SegmentElements.Where(x => x.Name == segmentElement.Name).Count() != 1)
                        throw new InvalidOperationException("There are no elements or there are multiple elements with the same name.");

                    // Check that the element has no other elements
                    if (segmentElement.Elements().Count() != 0)
                        throw new InvalidOperationException("Cannot get the element which has other elements.");

                    // Check if type has been specified inside the element
                    if (segmentElement.Attribute("Type") == null)
                        throw new InvalidOperationException("Type not specified.");

                    // Get element type
                    ElementType = GetElementType(segmentElement);

                    // Add value to the expando object
                    ExpandoDict.Add(segmentElement.Name.LocalName, Convert.ChangeType(segmentElement.Value, ElementType, CultureInfo.InvariantCulture));
                }

                return Segment;
            }
        }

        /// <summary>
        /// Sets the specified segment.
        /// </summary>
        /// <param name="segmentName">Segment name.</param>
        /// <param name="segmentAttribute">Segment attribute.</param>
        /// <param name="segment">Segment data.</param>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentException"></exception>
        public void SetSegment(string segmentName, string segmentAttribute, object segment)
        {
            // Check if segment name is null or empty
            if (string.IsNullOrEmpty(segmentName))
                throw new ArgumentException("Segment name is null or empty.");

            // Check if segment attribute is null or empty
            if (string.IsNullOrEmpty(segmentAttribute))
                throw new ArgumentException("Segment attribute is null or empty.");

            // Check if segment is null
            if (segment == null)
                throw new ArgumentException("Value is null.");

            // Lock object and set segment
            lock (ObjLock)
            {
                IEnumerable<XElement> SegmentElements;
                IDictionary<string, object> ExpandoDict = segment as IDictionary<string, object>;

                // Get segment elements
                SegmentElements = GetSegmentElements(segmentName, segmentAttribute);

                // Go through the properties of the expando object
                foreach (KeyValuePair<string, object> prop in ExpandoDict)
                {
                    XElement SegmentElement;
                    Type ElementType;

                    // Check that only one element with a given name exists
                    if (SegmentElements.Where(x => x.Name == prop.Key).Count() != 1)
                        throw new InvalidOperationException("There are no elements or there are multiple elements with the same name.");

                    // Get segment element
                    SegmentElement = GetSegmentElement(SegmentElements, prop.Key);

                    // Check if type has been specified inside the element
                    if (SegmentElement.Attribute("Type") == null)
                        throw new InvalidOperationException("Type not specified.");

                    // Get element type
                    ElementType = GetElementType(SegmentElement);

                    // Check if value type is valid
                    if (prop.Value.GetType() != ElementType)
                        throw new InvalidOperationException("Value type is invalid.");

                    // Update element with new value
                    SegmentElement.Value = Convert.ToString(prop.Value, CultureInfo.InvariantCulture);
                }

                // Save the xml
                XMLRoot.Save(Path.Combine(XMLPath, FileName));
            }
        }

        /// <summary>
        /// Get all available segments of a given name.
        /// </summary>
        /// <param name="segmentName">Segment name.</param>
        /// <param name="segmentAttribute">Segment attribute.</param>
        /// <param name="name">Specified segment.</param>
        /// <returns></returns>
        /// <exception cref="InvalidOperationException"></exception>
        /// <exception cref="ArgumentException"></exception>
        public List<object> GetSegments(string segmentName, string segmentAttribute, string name)
        {
            List<object> AvailableSegments = new List<object>();

            // Check if segment name is null or empty
            if (string.IsNullOrEmpty(segmentName))
                throw new ArgumentException("Segment name is null or empty.");

            // Check if segment attribute is null or empty
            if (string.IsNullOrEmpty(segmentAttribute))
                throw new ArgumentException("Segment attribute is null or empty.");

            // Check if name is null or empty
            if (string.IsNullOrEmpty(name))
                throw new ArgumentException("Name is null or empty.");

            // Lock and get segment
            lock (ObjLock)
            {
                IEnumerable<XElement> MainSegments;

                // Get main segments
                MainSegments = GetSegmentElements(segmentName, segmentAttribute);

                // Go through the segments
                foreach(XElement mainSegment in MainSegments)
                {
                    // Segment must have specified name
                    if(mainSegment.Name == name)
                    {
                        dynamic Segment = new ExpandoObject();
                        IDictionary<string, object> ExpandoDict = Segment as IDictionary<string, object>;
                        IEnumerable<XElement> SegmentElements = mainSegment.Elements();

                        // Go through the elements of the main segment
                        foreach(XElement segmentElement in SegmentElements)
                        {
                            Type ElementType;

                            // Check that only one element with a given name exists
                            if (SegmentElements.Where(x => x.Name == segmentElement.Name).Count() != 1)
                                throw new InvalidOperationException("There are no elements or there are multiple elements with the same name.");

                            // Check that the element has no other elements
                            if (segmentElement.Elements().Count() != 0)
                                throw new InvalidOperationException("Cannot get the element which has other elements.");

                            // Check if type has been specified inside the element
                            if (segmentElement.Attribute("Type") == null)
                                throw new InvalidOperationException("Type not specified.");

                            // Get element type
                            ElementType = GetElementType(segmentElement);

                            // Add value to the expando object
                            ExpandoDict.Add(segmentElement.Name.LocalName, Convert.ChangeType(segmentElement.Value, ElementType, CultureInfo.InvariantCulture));
                        }

                        // Add expando object to the list
                        AvailableSegments.Add(Segment);
                    }
                }
            }

            return AvailableSegments;
        }
        #endregion

        #region Private methods
        /// <summary>
        /// Gets the segment elements.
        /// </summary>
        /// <param name="segmentName">Segment name.</param>
        /// <param name="segmentAttribute">Segment attribute.</param>
        /// <returns>Segment elements.</returns>
        private IEnumerable<XElement> GetSegmentElements(string segmentName, string segmentAttribute)
        {
            IEnumerable<XElement> Segments;
            IEnumerable<XElement> SegmentElements;

            // Try to get the segments. Only one must exist
            Segments = XMLRoot.Elements(segmentName).Where(x => x.Attribute("Label")?.Value == segmentAttribute);

            if (Segments.Count() != 1)
                throw new InvalidOperationException("There is no segment or there are multiple segments with the same name and attribute.");

            // Get elements of a specified segment. Check if it has some
            SegmentElements = Segments.Elements();

            if (SegmentElements.Count() == 0)
                throw new InvalidOperationException("Cannot get the segment which does not have elements.");

            return SegmentElements;
        }

        /// <summary>
        /// Gets the segment element.
        /// </summary>
        /// <param name="segmentElements">Segment elements.</param>
        /// <param name="name">Element name.</param>
        /// <returns>Segment element.</returns>
        private XElement GetSegmentElement(IEnumerable<XElement> segmentElements, string name)
        {
            XElement SegmentElement;

            // Get the element
            SegmentElement = segmentElements.Where(x => x.Name == name).First();

            // Check that the element has no other elements
            if (SegmentElement.Elements().Count() != 0)
                throw new InvalidOperationException("Cannot get the element which has other elements.");

            return SegmentElement;
        }

        /// <summary>
        /// Gets the segment element type.
        /// </summary>
        /// <param name="segmentElement">Segment element.</param>
        /// <returns>Segment element type.</returns>
        private Type GetElementType(XElement segmentElement)
        {
            Type ElementType;

            // Create actual type from the element. Check if type is valid
            ElementType = Type.GetType(segmentElement.Attribute("Type").Value);

            if (ElementType == null)
                throw new InvalidOperationException("Element type is invalid.");

            return ElementType;
        }
        #endregion
    }
}

File: PLCSnap7Filters/XMLSettings/XMLSettingsManager.cpp
#include "XMLSettingsManager.h"
#include <filesystem>
#include <sstream>
#include <algorithm>

#include "../XMLSource/pugixml.hpp"

namespace fs = std::filesystem;

namespace XMLSettings
{

// ── Konstruktor ───────────────────────────────────────────────────────────────

XMLSettingsManager::XMLSettingsManager(const std::string& folderPath,
                                       const std::string& fileName)
{
    if (folderPath.empty())
        throw std::invalid_argument("Folder path is null or empty.");
    if (fileName.empty())
        throw std::invalid_argument("File name is null or empty.");

    XMLPath  = folderPath;
    FileName = fileName;

    fs::path fullPath = fs::path(folderPath) / fileName;
    pugi::xml_parse_result result = XMLDoc.load_file(fullPath.string().c_str());

    if (!result)
        throw std::runtime_error(std::string("XML parse error: ") + result.description());

    XMLRoot = XMLDoc.document_element();
}

// ── Javne metode ──────────────────────────────────────────────────────────────

XMLValue XMLSettingsManager::GetElement(const std::string& segmentName,
                                        const std::string& segmentAttribute,
                                        const std::string& name)
{
    if (segmentName.empty())
        throw std::invalid_argument("Segment name is null or empty.");
    if (segmentAttribute.empty())
        throw std::invalid_argument("Segment attribute is null or empty.");
    if (name.empty())
        throw std::invalid_argument("Name is null or empty.");

    std::lock_guard<std::mutex> lock(Mtx);

    auto elements = GetSegmentElements(segmentName, segmentAttribute);

    long count = std::count_if(elements.begin(), elements.end(),
        [&](const pugi::xml_node& n){ return std::string(n.name()) == name; });

    if (count != 1)
        throw std::runtime_error("There are no elements or there are multiple elements with the same name.");

    pugi::xml_node element = GetSegmentElement(elements, name);

    if (!element.attribute("Type"))
        throw std::runtime_error("Type not specified.");

    return ParseElementValue(element);
}

void XMLSettingsManager::SetElement(const std::string& segmentName,
                                    const std::string& segmentAttribute,
                                    const std::string& name,
                                    const XMLValue& value)
{
    if (segmentName.empty())
        throw std::invalid_argument("Segment name is null or empty.");
    if (segmentAttribute.empty())
        throw std::invalid_argument("Segment attribute is null or empty.");
    if (name.empty())
        throw std::invalid_argument("Name is null or empty.");

    std::lock_guard<std::mutex> lock(Mtx);

    auto elements = GetSegmentElements(segmentName, segmentAttribute);

    long count = std::count_if(elements.begin(), elements.end(),
        [&](const pugi::xml_node& n){ return std::string(n.name()) == name; });

    if (count != 1)
        throw std::runtime_error("There are no elements or there are multiple elements with the same name.");

    pugi::xml_node element = GetSegmentElement(elements, name);

    if (!element.attribute("Type"))
        throw std::runtime_error("Type not specified.");

    element.text().set(ValueToString(value).c_str());
    SaveXML();
}

std::map<std::string, XMLValue>
XMLSettingsManager::GetSegment(const std::string& segmentName,
                               const std::string& segmentAttribute)
{
    if (segmentName.empty())
        throw std::invalid_argument("Segment name is null or empty.");
    if (segmentAttribute.empty())
        throw std::invalid_argument("Segment attribute is null or empty.");

    std::lock_guard<std::mutex> lock(Mtx);

    auto elements = GetSegmentElements(segmentName, segmentAttribute);
    std::map<std::string, XMLValue> result;

    for (const pugi::xml_node& el : elements)
    {
        if (!el.attribute("Type"))
            throw std::runtime_error("Type not specified.");
        result[el.name()] = ParseElementValue(el);
    }

    return result;
}

void XMLSettingsManager::SetSegment(const std::string& segmentName,
                                    const std::string& segmentAttribute,
                                    const std::map<std::string, XMLValue>& segment)
{
    if (segmentName.empty())
        throw std::invalid_argument("Segment name is null or empty.");
    if (segmentAttribute.empty())
        throw std::invalid_argument("Segment attribute is null or empty.");

    std::lock_guard<std::mutex> lock(Mtx);

    auto elements = GetSegmentElements(segmentName, segmentAttribute);

    for (const auto& [key, val] : segment)
    {
        long count = std::count_if(elements.begin(), elements.end(),
            [&](const pugi::xml_node& n){ return std::string(n.name()) == key; });

        if (count != 1)
            throw std::runtime_error("There are no elements or there are multiple elements with the same name.");

        pugi::xml_node element = GetSegmentElement(elements, key);

        if (!element.attribute("Type"))
            throw std::runtime_error("Type not specified.");

        element.text().set(ValueToString(val).c_str());
    }

    SaveXML();
}

std::vector<std::map<std::string, XMLValue>>
XMLSettingsManager::GetSegments(const std::string& segmentName,
                                const std::string& segmentAttribute,
                                const std::string& name)
{
    if (segmentName.empty())
        throw std::invalid_argument("Segment name is null or empty.");
    if (segmentAttribute.empty())
        throw std::invalid_argument("Segment attribute is null or empty.");
    if (name.empty())
        throw std::invalid_argument("Name is null or empty.");

    std::lock_guard<std::mutex> lock(Mtx);

    auto mainElements = GetSegmentElements(segmentName, segmentAttribute);
    std::vector<std::map<std::string, XMLValue>> result;

    for (const pugi::xml_node& mainEl : mainElements)
    {
        if (std::string(mainEl.name()) != name)
            continue;

        std::map<std::string, XMLValue> segmentMap;

        for (pugi::xml_node child : mainEl.children())
        {
            // Preveri, da element nima podelementov
            if (child.first_child().type() == pugi::node_element)
                throw std::runtime_error("Cannot get the element which has other elements.");

            if (!child.attribute("Type"))
                throw std::runtime_error("Type not specified.");

            segmentMap[child.name()] = ParseElementValue(child);
        }

        result.push_back(std::move(segmentMap));
    }

    return result;
}

// ── Zasebne metode ────────────────────────────────────────────────────────────

std::vector<pugi::xml_node>
XMLSettingsManager::GetSegmentElements(const std::string& segmentName,
                                       const std::string& segmentAttribute)
{
    std::vector<pugi::xml_node> segments;

    for (pugi::xml_node child : XMLRoot.children(segmentName.c_str()))
    {
        pugi::xml_attribute labelAttr = child.attribute("Label");
        if (labelAttr && std::string(labelAttr.value()) == segmentAttribute)
            segments.push_back(child);
    }

    if (segments.size() != 1)
        throw std::runtime_error(
            "There is no segment or there are multiple segments with the same name and attribute.");

    std::vector<pugi::xml_node> elements;
    for (pugi::xml_node el : segments[0].children())
        elements.push_back(el);

    if (elements.empty())
        throw std::runtime_error("Cannot get the segment which does not have elements.");

    return elements;
}

pugi::xml_node
XMLSettingsManager::GetSegmentElement(const std::vector<pugi::xml_node>& segmentElements,
                                      const std::string& name)
{
    for (const pugi::xml_node& el : segmentElements)
    {
        if (std::string(el.name()) == name)
        {
            // Element ne sme imeti podelementov
            if (el.first_child().type() == pugi::node_element)
                throw std::runtime_error("Cannot get the element which has other elements.");
            return el;
        }
    }
    throw std::runtime_error("Element not found: " + name);
}

XMLValue XMLSettingsManager::ParseElementValue(const pugi::xml_node& node)
{
    std::string typeStr = node.attribute("Type").value();
    std::string valStr  = node.text().as_string();

    if (typeStr == "int" || typeStr == "System.Int32")
        return std::stoi(valStr);
    else if (typeStr == "double" || typeStr == "System.Double")
        return std::stod(valStr);
    else if (typeStr == "bool" || typeStr == "System.Boolean")
        return valStr == "true" || valStr == "True" || valStr == "1";
    else // string ali System.String
        return valStr;
}

std::string XMLSettingsManager::ValueToString(const XMLValue& value)
{
    return std::visit([](const auto& v) -> std::string {
        if constexpr (std::is_same_v<std::decay_t<decltype(v)>, bool>)
            return v ? "true" : "false";
        else if constexpr (std::is_same_v<std::decay_t<decltype(v)>, std::string>)
            return v;
        else
            return std::to_string(v);
    }, value);
}

void XMLSettingsManager::SaveXML()
{
    fs::path fullPath = fs::path(XMLPath) / FileName;
    if (!XMLDoc.save_file(fullPath.string().c_str()))
        throw std::runtime_error("Failed to save XML file: " + fullPath.string());
}

} // namespace XMLSettings
